### ISSUE TITLE
NaNs propagating to grad even when masked out

### ISSUE THREAD

**Issue by valtron** [pytorch/pytorch#64305]
## 🐛 Bug

I have a model with multiple outputs. Sometimes, certain outputs are missing, so they are NaN, and when calculating the loss they should be ignored. This can be done in two ways:

1) Mask out the missing outputs, then calculate the loss on the rest
2) Calculate the loss for each output (giving NaN for missing outputs) then mask out the NaNs

These should be equivalent, but they are not. The second can lead to NaNs in the gradient. See the example below.

## To Reproduce

```python
import torch

torch.manual_seed(0)

m = torch.nn.Linear(1, 2, bias = False)

x = torch.tensor([[0.0]])
y = torch.tensor([[float('nan'), 0.0]])
y_mask = ~torch.isnan(y)
y_pred = m(x)

def loss_fn(y_pred, y):
	# problem shows up when using l2 or log loss (but not l1)
	return torch.square(y_pred - y)

# these two branches should be equivalent; they aren't
if True: # (1) mask after loss_fn
	# doing it this way results in NaNs in the grad
	l = loss_fn(y_pred, y)[y_mask]
else:    # (2) mask before loss_fn
	l = loss_fn(y_pred[y_mask], y[y_mask])

# loss is fine: no NaNs
print("loss:", l)
assert not torch.any(torch.isnan(l))

l.sum().backward()

# grad should not have NaNs
assert not torch.any(torch.isnan(m.weight.grad))
```

## Expected behavior

The above code should not throw an assertion error, i.e. `m.weight.grad` should not have any NaNs.

## Environment

PyTorch version: 1.9.0+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro
GCC version: (Rev5, Built by MSYS2 project) 10.3.0
Clang version: 12.0.1
CMake version: version 3.21.2
Libc version: N/A

Python version: 3.9.1 (tags/v3.9.1:1e5d33e, Dec  7 2020, 17:08:21) [MSC v.1927 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.18362-SP0
Is CUDA available: True
CUDA runtime version: 11.2.152
GPU models and configuration: GPU 0: GeForce GTX 970
Nvidia driver version: 461.33
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] botorch==0.5.0
[pip3] gpytorch==1.5.0
[pip3] mypy==0.910
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.1+mkl
[pip3] torch==1.9.0+cu111
[pip3] torchaudio==0.9.0
[pip3] torchvision==0.10.0+cu111
[conda] Could not collect

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by SsnL** [comment]
This actually makes sense if you think about how backward mode autodiff works. Related: https://github.com/pytorch/pytorch/issues/52248

**Comment by albanD** [comment]
I am afraid this is expected behavior for now and it will be significant work to fix that.
We'll keep the discussion in #52248 and close this one for now.
